### PR TITLE
fiat-crypto powered Poly1305

### LINKED ISSUE #198: Implementing Poly1305 using fiat-crypto
As part of an investigation into implementing support for ECC in Orion, the fiat-crypto (https://github.com/mit-plv/fiat-crypto) has been used as a source for formally verified arithmetic operations. This greatly increases confidence in the correctness of the implementation, while also reducing the amount of work needed to maintain and audit implementations utilizing fiat-crypto.

The draft PR https://github.com/orion-rs/orion/pull/197 showcases these benefits well IMO. fiat-crypto also offers Rust generated code for Poly1305: https://github.com/mit-plv/fiat-crypto/blob/master/fiat-rust/src/poly1305_64.rs

I think it would be worth to explore replacing the current Poly1305 implementation with one that leverages the formally verified code generated by fiat-crypto, for the exact same reasons mentioned above.

### PR DESCRIPTION
**Pull request by brycx**
closes #198 

TODO:

- [x] Switch to fiat-crypto crate after #252 is merged
- [x] Ensure all `allow`s are good
- [x] Poly1305 is exposed from fiat-crypto crate, see https://github.com/mit-plv/fiat-crypto/issues/1089
- [x] Mention of formally-verified crypto in the README security section

### CODE REVIEW

File: src/hazardous/mac/poly1305.rs [hunk 1]
```diff
@@ -143,93 +148,29 @@ impl core::fmt::Debug for Poly1305 {
...
         }
 
-        let hibit: u32 = if self.is_finalized {
-            0
-        } else {
-            1 << 24
-        };
+        let mut mb = [0u8; 17];
+        mb[..16].copy_from_slice(data);
+        // One byte is appended to detect trailing zeroes if not last chunk.
```
Comment (vlmutolo) [inline src/hazardous/mac/poly1305.rs:164]:
```rust
// One byte is appended to detect trailing zeroes if not last chunk.
```

Is this from a Poly1305 standard somewhere? If it is, maybe we can point to a specific section from that standard in this comment. I think if something is surprising enough to warrant a comment, someone might want to go read through the spec for context.

File: src/hazardous/mac/poly1305.rs [hunk 2]
```diff
@@ -143,93 +148,29 @@ impl core::fmt::Debug for Poly1305 {
...
-
-        self.a[0] = h0;
-        self.a[1] = h1;
-        self.a[2] = h2;
-        self.a[3] = h3;
-        self.a[4] = h4;
+        // h += m
+        let mut h: fiat_poly1305_loose_field_element = [0u32; 5];
+        fiat_poly1305_add(&mut h, &self.a, &m);
+        // h *= r with partial reduction modulo p
```
Comment (vlmutolo) [inline src/hazardous/mac/poly1305.rs:172]:
```rust
// h += m
```

```rust
// h *= r with partial reduction modulo p
```

These were very helpful comments for my understanding. 👍🏼

**Review by vlmutolo — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -15,6 +15,8 @@ Currently supports:
 ### Security
 This library has **not undergone any third-party security audit**. Usage is at **own risk**.
 
+Orion uses formally-verified arithmetic, generated by Fiat Crypto, for the X25519 and Poly1305 implementations.
```
[the remark this answers is not in the document]
Comment (brycx) [reply README.md]: Thanks, fixed!

File: src/hazardous/mac/poly1305.rs (hunk 1 above)
Comment (brycx) [reply src/hazardous/mac/poly1305.rs:164]: Linked to the original Poly1305 design by DJB.

File: src/hazardous/mac/poly1305.rs (hunk 2 above)
Comment (brycx) [reply src/hazardous/mac/poly1305.rs:173]: Their actually leftovers from the Andrew Moon adaption. But yeah, I thought we'd just keep them there.

File: src/hazardous/mac/poly1305.rs (hunk 1 above)
Comment (vlmutolo) [reply src/hazardous/mac/poly1305.rs:164]: That's perfect 👍

**Review by vlmutolo — APPROVED** [review]
Looks great!

**Pull request merged**